File: 0006 Z 字形变换/main.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        floor = 1
        idx = 0
        cs = ""
        single = True
        i=0

        if(numRows==1):
            return s
        while i <len(s):
            i+=1
            if(idx < len(s)):
                # print(floor,idx,len(s),cs)
                if(floor == 1 or floor == numRows):
                    cs += s[idx]
                    idx += 2*numRows - 2
                else:
                    cs += s[idx]
                    if(single):
                        idx += 2*(numRows-floor)
                    else:
                        idx += 2*floor - 2
                    single = not single
            else:
                single = True
                idx = floor
                floor += 1
                i = i - 1
                # print(i)
        return cs
a=Solution()
print(a.convert("AB",1))

File: 0006 Z 字形变换/main.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows==1){
            return s;
        }
        string res="";
        int t=0;
        int idx=0;
        for(int i=0;i<numRows;i++){
            t=0;
            idx=i;
            if(i==0 || i== numRows -1){
                while( t* (2*numRows-2)+i <s.size()){
                    res.push_back(s[ t++ * (2*numRows-2) + i]);
                }
            }
            else{
                while( idx<s.size() ){
                    idx=t * (2*numRows-2) + i;
                    if(idx < s.size()){
                        res.push_back(s[idx]);
                    }
                    idx=++t * (2 * numRows-2) - i;
                    if(idx < s.size()){
                        res.push_back(s[idx]);
                    }
                }
            }
        }
        
        return res;
    }
};